crypto/data_formats_test.go

```diff
@@ -16,14 +16,93 @@ package crypto
 
 import (
 	"bytes"
+	"reflect"
 	"testing"
 
 	"github.com/google/trillian"
 	"github.com/google/trillian/testonly"
+	"github.com/kr/pretty"
 )
 
 var dh = testonly.MustHexDecode
 
+func TestLogRoot(t *testing.T) {
+	for _, logRoot := range []*LogRootV1{
+		{
+			Version:  1,
+			RootHash: []byte("foo"),
+			Metadata: []byte{},
+		},
+	} {
+		b, err := SerializeLogRoot(logRoot)
+		if err != nil {
+			t.Errorf("SerializeLogRoot(%v): %v", logRoot, err)
+		}
+		r2, err := ParseLogRoot(b)
+		if err != nil {
+			t.Errorf("ParseLogRoot(): %v", err)
+		}
+		if got, want := r2, logRoot; !reflect.DeepEqual(got, want) {
+			t.Errorf("serialize/parse round trip failed. got %#v, want %#v", pretty.Formatter(got), pretty.Formatter(want))
+		}
+	}
+}
+
+func TestParseLogRoot(t *testing.T) {
+	for _, tc := range []struct {
+		logRoot []byte
+		wantErr bool
+	}{
+		{
+			logRoot: func() []byte {
+				b, _ := SerializeLogRoot(&LogRootV1{})
+				return b
+			}(),
+		},
+		{
+			logRoot: func() []byte {
+				b, _ := SerializeLogRoot(&LogRootV1{})
+				b[0] = 1
```

Comment:
A comment would be helpful here.

Also/alternatively, it might be nice if the tests distinguished errors -- I think this one is checking that a wrong version (0x101) gets bounced (and so the error will be something containing "Invalid MapRoot.Version"), whereas the test below is a plain parse failure.

Reply:
It looks like the error types are internal to the `tls` package. It's difficult to reliably discern what the error was from the outside?

Reply:
I often do something like `if !strings.Contains(err.Error(), test.wantErr) {` and put a relevant substring in the test definition (`"failed to parse"`, `"trailing data"`, etc.) -- that seems to get a decent balance between specificity and brittleness.